### ISSUE TITLE
Possible bug in BBBlayers

### ISSUE THREAD

**Issue by homaralex** [kumar-shridhar/PyTorch-BayesianCNN#7]
It seems to me that there is a bug in `BBBlayers` - both in `BBBConv2d` and in `BBBLinearFactorial`.
You declare the class fields `self.conv_qw` and `self.fc_qw`, which are used for computing the KL divergence. The problem is, these parameters are not used with the actual data - in the `probforward` methods you are using `self.qw_mean` instead. Thus it looks as if you are actually not imposing the prior on the model weights (which are used for the tasks) but on a different set which is never used.
@kumar-shridhar

**Comment by homaralex** [comment]
Furthermore, you do the forward pass with `self.log_alpha`, but compute KL divergence with `self.qw_logvar`. These are all serious bugs

**Comment by kumar-shridhar** [comment]
Hi @homaralex ,
 in the probforward methods you are using self.qw_mean instead: You mean in the reparameterization trick, we use qw.mean to update conv_qw.mean ?
Can you do a pull request and update your suggestions there. It would be much clear to see things that way.

**Comment by homaralex** [comment]
In `probforward()` you do:

` conv_qw_mean = F.conv2d(input=input, weight=self.qw_mean, ...)`
`conv_qw_std = torch.sqrt(1e-8 + F.conv2d(input=input.pow(2), weight=torch.exp(self.log_alpha)*self.qw_mean.pow(2), ...))`

which will store gradients for `self.qw_mean` and `self.log_alpha`.

However, later when computing KL divergence you do:
`qw_logpdf = self.conv_qw.logpdf(w_sample)`
`kl = torch.sum(qw_logpdf - self.pw.logpdf(w_sample))`

So the gradients corresponding to KL divergence will be computed w.r.t. `self.conv_qw`.
Now, these variables (`self.conv_qw` and `self.qw`, `self.conv_qw_mean` and `self.qw_mean`, etc.) aren't connected in any way - you can see this by inspecting the model parameters (e.g. with a debugger). 

When I will have some free time I will try to pack this into a PR :)

**Comment by ShellingFord221** [comment]
Hi @homaralex , I use the code to separately train the weights with data and kl (just like your first comment in this issue), and the test results are surprisingly good. This is like not imposing the prior on the model weights (just like you said), and the model should easily overfit because it doesn't make a tradeoff between the likelihood cost and complexity cost. But why is the test result so good (the task is a NLP one)? And even the simpler the model is, the better results are. This is against my experience. I really wonder why. Thanks!

**Comment by kumar-shridhar** [comment]
Hi @ShellingFord221, What do you mean by training the two separately?

**Comment by ShellingFord221** [comment]
Hi, it's just your old code, where using qw to compute the data likelihood, conv_qw or fc_qw to compute kl, and these parameters are seperate (now I know that computing data cost and kl should using the same mean and variance of the weight, which means the balance between them). In this situation, when converging, the kl should be zero (i.e. conv_qw perfectly approximate the real distribution p) and qw perfectly fits the train data. I thought this situation will easily overfit when testing, but the test results are REALLY good. I'm very confused...

**Comment by kumar-shridhar** [comment]
I tried doing that but the network failed to converge easily. Would you mind doing a pull request and I will have a look too?
